kernel/src/driver_manager.rs

```diff
@@ -0,0 +1,156 @@
+
+use super::device_tree::DeviceTree;
+use super::globals;
+use super::mm;
+use super::paging::PagingImpl as _;
+use drivers::{Console, Driver};
+
+pub struct DriverManager {
+    drivers: LinkedList<Arc<dyn Driver>>,
+}
+
+impl DriverManager {
+    fn new() -> Self {
+        Self {
+            drivers: LinkedList::new(),
+        }
+    }
+
+    pub fn with_devices(dt: &DeviceTree) -> Self {
+        let mut mgr = Self::new();
+
+        mgr.do_console(dt).unwrap();
+
+        mgr
+    }
+
+    fn do_console(&mut self, dt: &DeviceTree) -> Result<(), &'static str> {
+        let cons_node = dt.console_node();
+        if cons_node.is_none() {
+            return Err("Device tree doesn't even contain a console node...");
+        }
+        let cons_node = cons_node.unwrap();
+
+        map_dt_regions(&cons_node);
+
+        if let Some(cons_driver) = self.find_console(&cons_node) {
+            self.register_console(cons_driver);
+            Ok(())
+        } else {
+            unmap_dt_regions(&cons_node);
+            Err("Failed to find a matching driver for the console...")
+        }
+    }
+
+    pub fn find_console(&self, cons_node: &fdt::node::FdtNode) -> Option<Box<dyn Console + Send + Sync>> {
+        let console_base = cons_node.reg()?.next()?.starting_address as usize;
+
+        // TODO: I'm not sure we are handling/parsing the compat string very well ^^
+        let compatible = cons_node
+            .properties()
+            .find(|prop| prop.name == "compatible")
+            .and_then(|some_prop| some_prop.as_str())
+            .unwrap_or("");
+        let compatibles = compatible.split('\0');
+
+
+        for compatible in compatibles {
+            if let Some(cons_driver) = drivers::matching_console_driver(compatible) {
+                return Some(cons_driver(console_base));
+            }
+        }
+
+        None
+    }
+
+    fn register_console(&mut self, cons: Box<dyn Console + Sync + Send>) {
+        let cons: Arc<dyn Console + Sync + Send> = Arc::from(cons);
+        self.register_driver(cons.clone());
+        globals::CONSOLE.set(cons.clone()).unwrap();
+    }
+
+    fn register_driver(&mut self, drv: Arc<dyn Driver>) {
+        self.drivers.push_back(drv);
+    }
+}
+
+pub fn matching_console_driver(
+    node: &fdt::node::FdtNode,
+) -> Option<Box<dyn Console + Sync + Send>> {
+    let compatible = node
+        .properties()
+        .find(|prop| prop.name == "compatible")
+        .and_then(|some_prop| some_prop.as_str())
+        .unwrap_or("");
+
+    let compatibles = compatible.split('\0');
+
+    let console_base = node.reg()?.next()?.starting_address as usize;
+
+    for compat in compatibles {
+        match compat {
+            "arm,pl011" => return Some(drivers::pl011::Pl011::boxed(console_base)),
+            // "ns16550" => return Some(&drivers::ns16550::Ns16550::new(console_base)),
+            _ => continue,
+        }
+    }
+
+    None
+}
```

Comment:
Is this still used with the new MATCHER mechanism?

Reply:
Thanks for the catch, this is dead code